src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs

```diff
@@ -0,0 +1,312 @@
+                {
+                    _endInstrumentMeasurements(instrument);
+                    RemoveInstrumentState(instrument, (InstrumentState)cookie!);
+                }
+            };
+            _listener.SetMeasurementEventCallback<double>((i, m, l, c) => ((InstrumentState)c!).Update((double)m, l));
+            _listener.SetMeasurementEventCallback<float>((i, m, l, c) => ((InstrumentState)c!).Update((double)m, l));
+            _listener.SetMeasurementEventCallback<long>((i, m, l, c) => ((InstrumentState)c!).Update((double)m, l));
+            _listener.SetMeasurementEventCallback<int>((i, m, l, c) => ((InstrumentState)c!).Update((double)m, l));
+            _listener.SetMeasurementEventCallback<short>((i, m, l, c) => ((InstrumentState)c!).Update((double)m, l));
+            _listener.SetMeasurementEventCallback<byte>((i, m, l, c) => ((InstrumentState)c!).Update((double)m, l));
+            _listener.SetMeasurementEventCallback<decimal>((i, m, l, c) => ((InstrumentState)c!).Update((double)m, l));
+        }
+
+
+
+        public AggregationManager Include(string meterName)
+        {
+            Include(i => i.Meter.Name == meterName);
+            return this;
+        }
+
+        public AggregationManager Include(string meterName, string instrumentName)
+        {
+            Include(i => i.Meter.Name == meterName && i.Name == instrumentName);
+            return this;
+        }
+
+        public AggregationManager Include(Predicate<Instrument> instrumentFilter)
+        {
+            lock (this)
+            {
+                _instrumentConfigFuncs.Add(instrumentFilter);
+            }
+            return this;
+        }
+
+        public AggregationManager SetCollectionPeriod(TimeSpan collectionPeriod)
+        {
+            lock (this)
+            {
+                _collectionPeriod = collectionPeriod;
+            }
+            return this;
+        }
+
+        public void Start()
+        {
+            // if already started or already stopped we can't be started again
+            if (_collectThread != null || _cts.IsCancellationRequested)
+            {
+                // correct usage from internal code should never get here
+                throw new InvalidOperationException("Start can only be called once");
+            }
+
+            if (_collectionPeriod.TotalSeconds < MinCollectionTime)
+            {
+                // correct usage from internal code should never get here
+                throw new InvalidOperationException($"CollectionPeriod must be >= {MinCollectionTime} sec");
+            }
+
+            // This explicitly uses a Thread and not a Task so that metrics still work
+            // even when an app is experiencing thread-pool starvation. Although we
+            // can't make in-proc metrics robust to everything, this is a common enough
+            // problem in .NET apps that it feels worthwhile to take the precaution.
```

Comment:
CC @stephentoub just in case he has any comment.

Reply:
How many of these AggregationManager instances will there be in a process, and thus how many of these threads?

Reply:
At any given time there should be at most one instance of AggregationManager and one thread. MetricsEventSource disposes the old one (which joins the thread) before creating new one.